File: src/BaseHmiTypes/Common/HmiMultilingualText.cs
using System.Globalization;
using System.Xml;
using System.Xml.Linq;

namespace BaseHmiTypes.Common;

public sealed class HmiMultilingualText
{
    public IDictionary<int, string> Texts { get; } = new Dictionary<int, string>();

    public IDictionary<int, string> FormattedTexts { get; } = new Dictionary<int, string>();

    public IDictionary<int, IList<HmiTextPart>> Parts { get; } = new Dictionary<int, IList<HmiTextPart>>();

    public int CategoryTypeId { get; set; }

    public string? CategorySubtype { get; set; }

    public static HmiMultilingualText FromText(string? text, int cultureLcid = -1)
    {
        var result = new HmiMultilingualText();
        result.Texts[cultureLcid] = text ?? string.Empty;
        return result;
    }

    public string GetText(CultureInfo? cultureInfo)
    {
        return GetCultureValue(Texts, cultureInfo) ?? GetDefaultText();
    }

    public string GetFormattedText(CultureInfo? cultureInfo)
    {
        return GetCultureValue(FormattedTexts, cultureInfo) ?? string.Empty;
    }

    public string GetDisplayText(CultureInfo? cultureInfo)
    {
        var formattedText = GetFormattedText(cultureInfo);
        if (!string.IsNullOrWhiteSpace(formattedText))
            return ExtractFormattedTextBody(formattedText);

        return GetText(cultureInfo);
    }

    public string GetFormattedTextBody(CultureInfo? cultureInfo)
    {
        return ExtractFormattedTextBody(GetFormattedText(cultureInfo));
    }

    public IReadOnlyList<HmiTextPart> GetParts(CultureInfo? cultureInfo)
    {
        if (Parts.Count == 0)
            return Array.Empty<HmiTextPart>();

        if (cultureInfo != null && Parts.TryGetValue(cultureInfo.LCID, out var cultureParts))
            return cultureParts as IReadOnlyList<HmiTextPart> ?? cultureParts.ToArray();

        if (Parts.TryGetValue(-1, out var invariantParts))
            return invariantParts as IReadOnlyList<HmiTextPart> ?? invariantParts.ToArray();

        if (Parts.TryGetValue(0, out var defaultParts))
            return defaultParts as IReadOnlyList<HmiTextPart> ?? defaultParts.ToArray();

        var first = Parts.First().Value;
        return first as IReadOnlyList<HmiTextPart> ?? first.ToArray();
    }

    public string GetDefaultText()
    {
        return Texts.Count == 0 ? string.Empty : Texts.First().Value ?? string.Empty;
    }

    public override string ToString()
    {
        return GetDefaultText();
    }

    private static string? GetCultureValue(IDictionary<int, string> values, CultureInfo? cultureInfo)
    {
        if (values.Count == 0)
            return null;

        if (cultureInfo != null && values.TryGetValue(cultureInfo.LCID, out var cultureValue))
            return cultureValue;

        if (values.TryGetValue(-1, out var invariantValue))
            return invariantValue;

        if (values.TryGetValue(0, out var defaultValue))
            return defaultValue;

        return values.First().Value ?? string.Empty;
    }

    private static string ExtractFormattedTextBody(string? formattedText)
    {
        if (string.IsNullOrWhiteSpace(formattedText))
            return string.Empty;

        try
        {
            using var reader = XmlReader.Create(
                new StringReader(formattedText!),
                new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit, XmlResolver = null });
            var document = XDocument.Load(reader, LoadOptions.PreserveWhitespace);
            var body = document
                .Descendants()
                .FirstOrDefault(element => string.Equals(element.Name.LocalName, "body", StringComparison.OrdinalIgnoreCase));

            if (body == null)
                return formattedText!;

            return string.Concat(body.Nodes().Select(node => RemoveNamespaces(node).ToString(SaveOptions.DisableFormatting)));
        }
        catch (XmlException)
        {
            return formattedText!;
        }
    }

    private static XNode RemoveNamespaces(XNode node)
    {
        if (node is XElement element)
        {
            return new XElement(
                element.Name.LocalName,
                element.Attributes()
                    .Where(attribute => !attribute.IsNamespaceDeclaration)
                    .Select(attribute => new XAttribute(attribute.Name.LocalName, attribute.Value)),
                element.Nodes().Select(RemoveNamespaces));
        }

        return node;
    }
}

File: typescript/base-hmi-types/src/common/HmiMultilingualText.ts
export class HmiMultilingualText {
  readonly texts = new Map<number, string>();
  readonly formattedTexts = new Map<number, string>();
  categoryTypeId = 0;
  categorySubtype?: string;

  static fromText(text: string | undefined, cultureId = -1): HmiMultilingualText {
    const result = new HmiMultilingualText();
    result.texts.set(cultureId, text ?? "");
    return result;
  }

  getText(cultureId?: number): string {
    return this.getCultureValue(this.texts, cultureId) ?? this.getDefaultText();
  }

  getFormattedText(cultureId?: number): string {
    return this.getCultureValue(this.formattedTexts, cultureId) ?? "";
  }

  getDisplayText(cultureId?: number): string {
    const formattedText = this.getFormattedText(cultureId);
    return formattedText.trim() === "" ? this.getText(cultureId) : HmiMultilingualText.extractFormattedTextBody(formattedText);
  }

  getFormattedTextBody(cultureId?: number): string {
    return HmiMultilingualText.extractFormattedTextBody(this.getFormattedText(cultureId));
  }

  getDefaultText(): string {
    for (const value of this.texts.values()) {
      return value ?? "";
    }

    return "";
  }

  toString(): string {
    return this.getDefaultText();
  }

  private getCultureValue(values: Map<number, string>, cultureId?: number): string | undefined {
    if (values.size === 0) {
      return undefined;
    }

    if (cultureId !== undefined && values.has(cultureId)) {
      return values.get(cultureId);
    }

    return values.get(-1) ?? values.get(0) ?? values.values().next().value ?? "";
  }

  private static extractFormattedTextBody(formattedText: string | undefined): string {
    if (formattedText === undefined || formattedText.trim() === "") {
      return "";
    }

    const match = /<body(?:\s[^>]*)?>([\s\S]*?)<\/body>/i.exec(formattedText);
    if (match === null) {
      return formattedText;
    }

    return match[1].replace(/\s+xmlns(?::\w+)?="[^"]*"/g, "");
  }
}
